### PR TITLE
Update to pydantic 2

### LINKED ISSUE #4: pydantic 2.0
tracking issue for adopting pydantic 2.0

### PR DESCRIPTION
**Pull request by clbarnes**
Resolves #4

- Use new-fangled syntax
  - depend on typing-extensions for py<3.12
- Change error messages in doctests
- Bump black, ruff dependencies for compatibility with pytest-examples
- Add black and ruff to dev dependencies (fixed versions)

### CODE REVIEW

File: src/pydantic_zarr/core.py [hunk 1]
```diff
@@ -137,7 +136,9 @@ def from_zarr(cls, zarray: zarr.Array):
             shape=zarray.shape,
             chunks=zarray.chunks,
             dtype=str(zarray.dtype),
-            fill_value=zarray.fill_value,
+            # explicitly cast to numpy type and back to python
+            # so that int 0 isn't serialized as 0.0
+            fill_value=zarray.dtype.type(zarray.fill_value).tolist(),
```
Comment (clbarnes) [inline src/pydantic_zarr/core.py:141]: This is worth noting

Comment (d-v-b) [reply src/pydantic_zarr/core.py:141]: good catch!

**Comment by d-v-b** [comment]
@clbarnes I pushed a lot of changes into this branch. perhaps the most important is the addition of [this test](https://github.com/clbarnes/pydantic-zarr/blob/2631c4c7fb38ae812fc1c4f97906bb0679eb9a57/tests/test_v2.py#L190) which is currently failing. Pydantic 2 [changed how it handles unions](https://docs.pydantic.dev/2.3/usage/types/unions/#union-mode), and for some reason it's not doing the right thing here -- a `ArraySpec` in dict form gets parsed as a `GroupSpec`, when it really shouldn't. Do you have any ideas for sorting this out? I don't think we can use `Annotated` here, or at least when I tried to use `Annotated` around a `TypeVar` used in generics I got some angry errors.

**Comment by d-v-b** [comment]
OK, this was my mistake, and it's now fixed (by re-adding the "no extra fields" constraint on the base model for array and group specs)

**Comment by d-v-b** [comment]
I'm can't understand why [this test](https://github.com/clbarnes/pydantic-zarr/blob/398aec41b3e2c701a0099a8f938c675f4cd48c8e/tests/test_v2.py#L133) is failing. Any ideas @clbarnes ?

**Pull request merged**